bftengine/src/bcstatetransfer/AsyncStateTransferCRE.cpp

```diff
@@ -136,8 +129,41 @@ class ScalingReplicaHandler : public IStateHandler {
     return logger_;
   }
 };
-std::shared_ptr<ClientReconfigurationEngine> CreFactory::create(std::shared_ptr<MsgsCommunicator> msgsCommunicator,
-                                                                std::shared_ptr<MsgHandlersRegistrator> msgHandlers) {
+
+// TODO(yf): remove
+class MainKeyUpdateHandler : public IStateHandler {
+ public:
+  MainKeyUpdateHandler() { LOG_INFO(getLogger(), "Created StateTransfer CRE replica main key update handler"); }
+  bool validate(const State& state) const override {
+    concord::messages::ClientStateReply crep;
+    concord::messages::deserialize(state.data, crep);
+    if (std::holds_alternative<concord::messages::ReplicaMainKeyUpdate>(crep.response)) {
+      concord::messages::ReplicaMainKeyUpdate update = std::get<concord::messages::ReplicaMainKeyUpdate>(crep.response);
+      return true;
+    }
+    return false;
+  }
+
+  bool execute(const State& state, WriteState&) override {
+    LOG_INFO(getLogger(), "execute MainKeyUpdateHandler");
+    concord::messages::ClientStateReply crep;
+    concord::messages::deserialize(state.data, crep);
+    concord::messages::ReplicaMainKeyUpdate update = std::get<concord::messages::ReplicaMainKeyUpdate>(crep.response);
+    bftEngine::CryptoManager::instance().onPublicKeyExchange(update.key, update.sender_id, update.seq_num);
```

Comment:
Is there a way to get the latest sender_id key's sequence number?
I wonder if it is possible already now to handle the following scenario:
The ROR has the main key of replica X for sequence number S1. However, as it reads the state only from f+1 replicas, then f Byzantine replicas + 1 late replica may give it X's key for sequence number S2 < S1.
If it is possible to get from Cryptomanager X's latest key sequence number, then it can be solved at the handler level.

On the other hand, we may need to handle globally, for all AsyncStateTransferCRE handlers...
WDYT?

Reply:
Lets handle it globally in a different PR